src/main/java/org/javarosa/xform/parse/XFormParser.java

```diff
@@ -480,25 +480,27 @@ private void parseDoc(Map<String, String> namespacePrefixesByUri, String lastSav
                 final String instanceId = instanceNodeIdStrs.get(instanceIndex);
                 final String instanceSrc = parseInstanceSrc(instance, lastSavedSrc);
 
-                // Only read in an external instance if its ID is used in the primary instance as an argument to an
+                // Only read in an secondary instance if its ID is used in the primary instance as an argument to an
                 // instance() call
-                if (instanceSrc != null && referencedInstanceIds.contains(instanceId)) {
-                    final ExternalDataInstance externalDataInstance;
-                    try {
-                        externalDataInstance = ExternalDataInstance.build(instanceSrc, instanceId);
-                    } catch (IOException | UnfullfilledRequirementsException | InvalidStructureException |
+                if (referencedInstanceIds.contains(instanceId)) {
+                    if (instanceSrc != null) {
+                        final ExternalDataInstance externalDataInstance;
```

Comment:
This is much more clear now: This branch deals with external instances, the next with internals.

We could take the chance to give names to the `instanceSrc != null` condition to reflect this.